x/dclupgrade/handler_test.go

```diff
@@ -0,0 +1,427 @@
+	// propose the same upgrade
+	_, err = setup.Handler(setup.Ctx, msgProposeUpgrade)
+	require.Error(t, err)
+	require.True(t, types.ErrProposedUpgradeAlreadyExists.Is(err))
+}
+
+func TestHandler_OnlyTrusteeCanProposeUpgrade(t *testing.T) {
+	setup := Setup(t)
+
+	trusteeAccAddress1 := testdata.GenerateAccAddress()
+	trusteeAccAddress2 := testdata.GenerateAccAddress()
+	trusteeAccAddress3 := testdata.GenerateAccAddress()
+	setup.AddAccount(trusteeAccAddress1, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress2, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress3, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+
+	setup.DclauthKeeper.On("CountAccountsWithRole", mock.Anything, dclauthtypes.Trustee).Return(3)
+
+	for _, role := range []dclauthtypes.AccountRole{
+		dclauthtypes.TestHouse,
+		dclauthtypes.CertificationCenter,
+		dclauthtypes.Vendor,
+		dclauthtypes.NodeAdmin,
+	} {
+		accAddress := testdata.GenerateAccAddress()
+		setup.AddAccount(accAddress, []dclauthtypes.AccountRole{role})
+
+		// propose upgrade by user without Trustee role
+		msgProposeUpgrade := NewMsgProposeUpgrade(accAddress)
+		_, err := setup.Handler(setup.Ctx, msgProposeUpgrade)
+		require.Error(t, err)
+		require.True(t, sdkerrors.ErrUnauthorized.Is(err))
+	}
+
+	// propose upgrade by user with Trustee role
+	msgProposeUpgrade := NewMsgProposeUpgrade(trusteeAccAddress1)
+
+	setup.UpgradeKeeper.On("ScheduleUpgrade", mock.Anything, msgProposeUpgrade.Plan).Return(nil)
+	_, err := setup.Handler(setup.Ctx, msgProposeUpgrade)
+	require.NoError(t, err)
+}
+
+func TestHandler_ProposeUpgradeWhenSeveralVotesNeeded(t *testing.T) {
+	setup := Setup(t)
+
+	trusteeAccAddress1 := testdata.GenerateAccAddress()
+	trusteeAccAddress2 := testdata.GenerateAccAddress()
+	trusteeAccAddress3 := testdata.GenerateAccAddress()
+	setup.AddAccount(trusteeAccAddress1, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress2, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+	setup.AddAccount(trusteeAccAddress3, []dclauthtypes.AccountRole{dclauthtypes.Trustee})
+
+	setup.DclauthKeeper.On("CountAccountsWithRole", mock.Anything, dclauthtypes.Trustee).Return(3)
+
+	// propose upgrade
+	msgProposeUpgrade := NewMsgProposeUpgrade(trusteeAccAddress1)
+
+	setup.UpgradeKeeper.On("ScheduleUpgrade", mock.Anything, msgProposeUpgrade.Plan).Return(nil)
+	_, err := setup.Handler(setup.Ctx, msgProposeUpgrade)
+	require.NoError(t, err)
+
+	setup.UpgradeKeeper.AssertCalled(
+		t,
+		"ScheduleUpgrade",
+		mock.MatchedBy(isContextWithCachedMultiStore),
+		msgProposeUpgrade.Plan,
+	)
+
+	setup.UpgradeKeeper.AssertNotCalled(
+		t,
+		"ScheduleUpgrade",
+		mock.MatchedBy(isContextWithNonCachedMultiStore),
+		msgProposeUpgrade.Plan,
+	)
+
+	// check proposed upgrade for being created
+	_, isFound := setup.Keeper.GetProposedUpgrade(setup.Ctx, msgProposeUpgrade.Plan.Name)
+	require.True(t, isFound)
```

Comment:
It makes sense to also verify the content of `ProposedUpgrade` entity.

Reply:
added